### PR TITLE
Update virtual documents conditionally

### PR DESCRIPTION
**Pull request by trungleduc**
## References

Partially addresses https://github.com/jupyterlab/jupyterlab/issues/15144

> The LSP virtual document is not listening to updates if there is no LSP feature registered or no server connected

## Code changes

- The virtual document is connected to the `contentChanged` signal if three conditions are satisfied: 
  -  The LSP feature is enabled.
  -  There is a connected server for the language of the document.
  - The LSP features list is not empty.

### CODE REVIEW

File: packages/lsp/src/adapters/adapter.ts [hunk 1]
```diff
@@ -718,4 +733,41 @@ export abstract class WidgetLSPAdapter<
...
+    if (!this._virtualDocument) {
+      return;
+    }
+
+    const languages = [
+      this._virtualDocument.language,
+      ...[...this._virtualDocument.foreignDocuments.values()].map(
+        it => it.language
+      )
+    ];
```
Comment (krassowski) [inline packages/lsp/src/adapters/adapter.ts:763]: I am afraid that this is a chicken and egg problem. Downstream we do not know all the languages until we updated the document as this is when the extraction happens.

**Review by krassowski — COMMENTED** [review]
- The LSP feature is enabled.
:+1: 
- The LSP features list is not empty.
:+1: 
- There is a connected server for the language of the document.
  This one is problematic when transclusions/nested documents are at play. We do not know the languages until we parsed (updated) the document. Is my reasoning correct?

Maybe for now it would be good enough to use the first two as good enough heuristic?

**Comment by trungleduc** [comment]
Failed UI tests do not seem to be related.

File: packages/lsp/test/adapters.spec.ts [hunk 2]
```diff
@@ -212,6 +214,130 @@ describe('@jupyterlab/lsp', () => {
       tracker.dispose();
     });
   });
+
+  describe('WidgetLSPAdapter1', () => {
```
Comment (krassowski) [inline packages/lsp/test/adapters.spec.ts:218]:
```suggestion
  describe('WidgetLSPAdapter', () => {
```

**Review by krassowski — APPROVED** [review]
Thank you @trungleduc!

File: packages/lsp/test/adapters.spec.ts (hunk 2 above)
Comment (trungleduc) [reply packages/lsp/test/adapters.spec.ts:218]: Thanks!

**Comment by krassowski** [comment]
Visual test failure unrelated, merging.

**Pull request merged**